File: android/src/main/java/io/dolby/sdk/reactnative/services/RNCommandServiceModule.java
package io.dolby.sdk.reactnative.services;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.voxeet.sdk.services.CommandService;
import com.voxeet.sdk.services.ConferenceService;

import org.jetbrains.annotations.NotNull;

/**
 * The RNCommandServiceModule allows the application to send {@link #send(String, Promise)} text messages to all other participants of
 * a specific conference.
 */
public class RNCommandServiceModule extends ReactContextBaseJavaModule {

    @NotNull
    private final ConferenceService conferenceService;
    @NotNull
    private final CommandService commandService;

    /**
     * Creates a bridge wrapper for {@link CommandService}.
     *
     * @param conferenceService {@link ConferenceService} from Android SDK
     * @param commandService    {@link CommandService} from Android SDK
     * @param reactContext      react context
     */
    public RNCommandServiceModule(
            @NotNull ConferenceService conferenceService,
            @NotNull CommandService commandService,
            @NotNull ReactApplicationContext reactContext
    ) {
        super(reactContext);
        this.conferenceService = conferenceService;
        this.commandService = commandService;
    }

    @NotNull
    @Override
    public String getName() {
        return "DolbyIoIAPICommandServiceModule";
    }

    /**
     * Sends the message to the conference. The message must be in the form of a string or a representation of strings (json or
     * base64).
     *
     * @param message content of the message (any possible string)
     * @param promise returns true if message was send, false otherwise
     */
    @ReactMethod
    public void send(@NotNull final String message, @NotNull Promise promise) {
        String conferenceId = conferenceService.getConferenceId();
        if (conferenceId == null) {
            promise.reject(new Exception("Couldn't find the conference"));
            return;
        }
        commandService.send(conferenceId, message)
                .then(promise::resolve)
                .error(promise::reject);
    }
}

File: android/src/main/java/io/dolby/sdk/reactnative/services/RNCommandServiceModule.kt
package io.dolby.sdk.reactnative.services

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.voxeet.sdk.services.CommandService
import com.voxeet.sdk.services.ConferenceService
import io.dolby.sdk.reactnative.eventemitters.RNCommandEventEmitter
import io.dolby.sdk.reactnative.eventemitters.RNEventEmitter
import io.dolby.sdk.reactnative.utils.Promises
import io.dolby.sdk.reactnative.utils.Promises.forward
import io.dolby.sdk.reactnative.utils.Promises.thenValue

/**
 * The [RNCommandServiceModule] allows the application to send [send] text messages to all other participants of
 * a specific conference.
 *
 * @constructor
 * Creates a bridge wrapper for [CommandService].
 *
 * @param reactContext      react context
 * @param conferenceService [ConferenceService] from Android SDK
 * @param commandService    [CommandService] from Android SDK
 * @param eventEmitter      an emitter for the command module events
 */
class RNCommandServiceModule(
  reactContext: ReactApplicationContext,
  eventEmitter: RNCommandEventEmitter,
  private val conferenceService: ConferenceService,
  private val commandService: CommandService
) : RNEventEmitterModule(reactContext, eventEmitter) {

  override fun getName(): String = "DolbyIoIAPICommandServiceModule"

  /**
   * Sends the message to the conference. The message must be in the form of a string or a representation of strings (json or
   * base64).
   *
   * @param message content of the message (any possible string)
   * @param promise returns true if message was send, false otherwise
   */
  @ReactMethod
  fun send(message: String, promise: Promise) {
    Promises.promise(conferenceService.conferenceId) { "Couldn't find the conference" }
      .thenValue { conferenceId -> commandService.send(conferenceId, message) }
      .forward(promise)
  }

  /**
   * Every emitter module must implement this method in place, otherwise JS cannot receive event
   */
  @ReactMethod
  override fun addListener(eventName: String) {
    super.addListener(eventName)
  }

  /**
   * Every emitter module must implement this method in place, otherwise JS cannot receive event
   */
  @ReactMethod
  override fun removeListeners(count: Int) {
    super.removeListeners(count)
  }

}
